### PR TITLE
Add IPv4 and IPv6 endpoint abstraction

### PR DESCRIPTION
**Pull request by jakobod**
This PR adds ipv4 and ipv6 endpoint abstractions to caf.

### CODE REVIEW

File: libcaf_core/src/ipv4_endpoint.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+size_t ipv4_endpoint::hash_code() const noexcept {
+  auto result = fnv_hash(address_.data());
+  return fnv_hash_append(result, port_);
+}
+
+long ipv4_endpoint::compare(ipv4_endpoint x) const noexcept {
+  auto res = address_.compare(x.address());
+  if (res != 0)
```
[the remark this answers is not in the document]
Comment (jakobod) [reply libcaf_core/src/ipv4_endpoint.cpp]: This should've been `if (res == 0)`..

File: libcaf_core/test/ipv4_endpoint.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  // First endpoint is always smaller than the second.
+  std::vector<comparison_testcase> comparison_testdata;
+  std::vector<to_string_testcase> to_string_testdata;
+
+private:
+  void add(endpoint ep1, endpoint ep2) {
+    comparison_testdata.emplace_back(comparison_testcase{ep1, ep2});
+  }
+
+  void add(endpoint ep, const std::string& str) {
```
Comment (Neverlord) [inline libcaf_core/test/ipv4_endpoint.cpp]: Having two functions named `add` that write into different vectors is hard to follow. Ideally, we simply initialize the vectors with the data instead of filling them piece by piece. What do you think of this strawman to make the code a bite more readable?


```c++
namespace {

ipv4_endpoint operator"" _ep(const char* str,size_t) {
  std::array<uint8_t, 4> bytes;
  char* next = nullptr;
  bytes[0] = static_cast<uint8_t>(strtol(str, &next, 10));
  for (size_t n = 1; n < 4; ++n) {
    assert(*next == '.');
    str = next + 1;
    bytes[n] = static_cast<uint8_t>(strtol(str, &next, 10));
  }
  assert(*next == ':');
  auto port = static_cast<uint16_t>(strtol(next + 1, nullptr, 10));
  return ipv4_endpoint{ipv4_address{bytes}, port};
}

class fixture {
  actor_system_config cfg;
  actor_system sys{cfg};
  std::vector<std::pair<ipv4_endpoint, ipv4_endpiont>> xs{
    {"127.0.0.1:8888"_ep, "127.0.0.2:8888"_ep},
    // ...
  };
};

#define CHECK_TO_STRING(addr)                                                  \
  CAF_CHECK_EQUAL(to_string(addr), addr##_ep)

} // namespace
```

This user-defined literal is of course not "production ready", but it's good enough as a unit test DSL. Now, you can also write a test for `to_string` simply like this:

```
CAF_TEST(to string) {
  CHECK_TO_STRING("127.0.0.1:8080");
  // ...
}
```

This is what the IPv6 version could look like:

```c++
ipv6_endpoint operator"" _ep(const char* str,size_t) {
  union {
    std::array<uint8_t, 16> xs_bytes;
    std::array<uint16_t, 8> xs;
  };
  union {
    std::array<uint8_t, 16> ys_bytes;
    std::array<uint16_t, 8> ys;
  };
  memset(xs_bytes.data(), 0, xs_bytes.size());
  memset(ys_bytes.data(), 0, xs_bytes.size());
  assert(*str == '[');
  ++str;
  char* next = nullptr;
  auto pull = [&] {
    auto x = static_cast<uint16_t>(strtol(str, &next, 16));
    return detail::to_network_order(x);
  };
  size_t n = 0;
  do {
    xs[n++] = pull();
    assert(next != nullptr);
    str = next + 1;
  } while (*str != ':' && *str != ']');
  if (*str == ':') {
    ++str;
    do {
      ys[0] = pull();
      std::rotate(ys.begin(), ys.begin() + 1, ys.end());
      assert(next != nullptr);
      str = next + 1;
    } while (*next != ']');
  }
  assert(*next == ']');
  assert(*str == ':');
  ++str;
  auto port = static_cast<uint16_t>(strtol(str, nullptr, 10));
  std::array<uint8_t, 16> bytes;
  for (size_t i = 0; i < 16; ++i)
    bytes[i] = xs_bytes[i] | ys_bytes[i];
  return ipv6_endpoint{ipv6_address{bytes}, port};
}
```

Btw, PR #882 adds new `parse` functions. Once that got merged, we can streamline the user-defined literals (but I think they're OK for now to make the unit tests more comprehensible).

File: libcaf_core/test/ipv4_endpoint.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      CAF_FAIL("serialization failed: " << sys.render(err));
+    binary_deserializer source(sys, make_span(buf));
+    ipv4_endpoint deserialized_data;
+    if (auto err = source(deserialized_data))
+      CAF_FAIL("deserialization failed: " << sys.render(err));
+    CAF_CHECK_EQUAL(testcase.first, deserialized_data);
+  }
+}
+
+CAF_TEST_FIXTURE_SCOPE_END()
```
Comment (Neverlord) [inline libcaf_core/test/ipv4_endpoint.cpp]: Missing newline at end of file.

File: libcaf_core/src/ipv6_endpoint.cpp [hunk 4]
```diff
@@ -20,47 +20,25 @@
...
-  if (res != 0)
-    return port_ - x.port();
-  else
-    return res;
+  return res == 0 ? port_ - x.port() : res;
 }
 
 std::string to_string(const ipv6_endpoint& ep) {
-  return to_string(ep.address()) + ":" + std::to_string(ep.port());
+  return "[" + to_string(ep.address()) + "]:" + std::to_string(ep.port());
```
[the remark this answers is not in the document]
Comment (jakobod) [reply libcaf_core/src/ipv6_endpoint.cpp]: I forgot about the embedding part. Now the proposed code makes sense! :)

File: libcaf_core/test/ipv6_endpoint.cpp [hunk 5]
```diff
@@ -138,45 +138,47 @@ CAF_TEST(constructing assigning and hash_code) {
...
+  CHECK_SERIALIZATION("[4f::1]:2222");
+  CHECK_SERIALIZATION("[4432:8d::33:1]:8732");
+  CHECK_SERIALIZATION("[4e::d00:0:ed00:0:1]:5678");
+  CHECK_SERIALIZATION("[::1]:2221");
+  CHECK_SERIALIZATION("[::1]:2222");
+  CHECK_SERIALIZATION("[4432::33:1]:872");
+  CHECK_SERIALIZATION("[4432::33:1]:999");
 }
 
 CAF_TEST_FIXTURE_SCOPE_END()
```
Comment (Neverlord) [inline libcaf_core/test/ipv6_endpoint.cpp]: Missing newline at end of file.

Comment (jakobod) [reply libcaf_core/test/ipv6_endpoint.cpp]: Dang.. I missed that one. :smile:

File: libcaf_core/caf/ipv4_endpoint.hpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+#pragma once
+
+#include "caf/detail/comparable.hpp"
+#include "caf/ipv4_address.hpp"
+#include "caf/meta/type_name.hpp"
+
+namespace caf {
+
+/// An IP endpoint that contains an ::ipv4_address and a port.
+struct ipv4_endpoint : detail::comparable<ipv4_endpoint> {
```
Comment (Neverlord) [inline libcaf_core/caf/ipv4_endpoint.hpp]:
```suggestion
class ipv4_endpoint : detail::comparable<ipv4_endpoint> {
```

File: libcaf_core/caf/ipv6_endpoint.hpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+#pragma once
+
+#include "caf/detail/comparable.hpp"
+#include "caf/ipv6_address.hpp"
+#include "caf/meta/type_name.hpp"
+
+namespace caf {
+
+/// An IP endpoint that contains an ::ipv6_address and a port.
+struct ipv6_endpoint : detail::comparable<ipv6_endpoint> {
```
Comment (Neverlord) [inline libcaf_core/caf/ipv6_endpoint.hpp]:
```suggestion
class ipv6_endpoint : detail::comparable<ipv6_endpoint> {
```
